### ISSUE TITLE
python310Packages.poetry is broken on unstable

### ISSUE THREAD

**Issue by RaitoBezarius** [NixOS/nixpkgs#204494]
### Steps To Reproduce
1. build `python310Packages.poetry`

### Build log
```
error: build of '/nix/store/10psvwb5fqcmcdnwfnx2h64jpn5xsbfi-python3.10-poetry-1.2.2.drv' on 'ssh://nix-community-build01' failed: builder for '/nix/store/10psvwb5fqcmcdnwfnx2h64jpn5xsbfi-python3.10-poetry-1.2.2.drv' failed with exit code 1;
       last 10 log lines:
       > Requirement already satisfied: dulwich<0.21.0,>=0.20.46 in /nix/store/1dwv35r9nr6x7hh9s0qfzfijn0ylc44n-python3.10-dulwich-0.20.50/lib/python3.10/site-packages (from poetry==1.2.2) (0.20.50)
       > Requirement already satisfied: html5lib<2.0,>=1.0 in /nix/store/zbi87b6967x988mdriihy6lm195xbib3-python3.10-html5lib-1.1/lib/python3.10/site-packages (from poetry==1.2.2) (1.1)
       > Requirement already satisfied: requests-toolbelt<0.10.0,>=0.9.1 in /nix/store/g6klr9lgfrkxf7wkdsf7r9pc39n45hgd-python3.10-requests-toolbelt-0.9.1/lib/python3.10/site-packages (from poetry==1.2.2) (0.9.1)
       > Requirement already satisfied: crashtest in /nix/store/m1sqz36r1zyjybi9321av7pyafvm9ql7-python3.10-crashtest-0.4.1/lib/python3.10/site-packages (from poetry==1.2.2) (0.4.1)
       > Requirement already satisfied: pexpect<5.0.0,>=4.7.0 in /nix/store/lxinahd9qhdzmqia3493mrgkqh3ingig-python3.10-pexpect-4.8.0/lib/python3.10/site-packages (from poetry==1.2.2) (4.8.0)
       > Requirement already satisfied: poetry-core==1.3.2 in /nix/store/y7m62x2v20cxga1vpqr7kmbgc6nibs5s-python3.10-poetry-core-1.3.2/lib/python3.10/site-packages (from poetry==1.2.2) (1.3.2)
       > Requirement already satisfied: urllib3<2.0.0,>=1.26.0 in /nix/store/fbac0h6j5har9447am8mqswnf1b2mrhh-python3.10-urllib3-1.26.12/lib/python3.10/site-packages (from poetry==1.2.2) (1.26.12)
       > ERROR: Could not find a version that satisfies the requirement cleo<2.0.0,>=1.0.0a5 (from poetry) (from versions: none)
       > ERROR: No matching distribution found for cleo<2.0.0,>=1.0.0a5
       > 
       For full logs, run 'nix log /nix/store/10psvwb5fqcmcdnwfnx2h64jpn5xsbfi-python3.10-poetry-1.2.2.drv'.
error: builder for '/nix/store/10psvwb5fqcmcdnwfnx2h64jpn5xsbfi-python3.10-poetry-1.2.2.drv' failed with exit code 1;
       last 10 log lines:
       > Requirement already satisfied: dulwich<0.21.0,>=0.20.46 in /nix/store/1dwv35r9nr6x7hh9s0qfzfijn0ylc44n-python3.10-dulwich-0.20.50/lib/python3.10/site-packages (from poetry==1.2.2) (0.20.50)
       > Requirement already satisfied: html5lib<2.0,>=1.0 in /nix/store/zbi87b6967x988mdriihy6lm195xbib3-python3.10-html5lib-1.1/lib/python3.10/site-packages (from poetry==1.2.2) (1.1)
       > Requirement already satisfied: requests-toolbelt<0.10.0,>=0.9.1 in /nix/store/g6klr9lgfrkxf7wkdsf7r9pc39n45hgd-python3.10-requests-toolbelt-0.9.1/lib/python3.10/site-packages (from poetry==1.2.2) (0.9.1)
       > Requirement already satisfied: crashtest in /nix/store/m1sqz36r1zyjybi9321av7pyafvm9ql7-python3.10-crashtest-0.4.1/lib/python3.10/site-packages (from poetry==1.2.2) (0.4.1)
       > Requirement already satisfied: pexpect<5.0.0,>=4.7.0 in /nix/store/lxinahd9qhdzmqia3493mrgkqh3ingig-python3.10-pexpect-4.8.0/lib/python3.10/site-packages (from poetry==1.2.2) (4.8.0)
       > Requirement already satisfied: poetry-core==1.3.2 in /nix/store/y7m62x2v20cxga1vpqr7kmbgc6nibs5s-python3.10-poetry-core-1.3.2/lib/python3.10/site-packages (from poetry==1.2.2) (1.3.2)
       > Requirement already satisfied: urllib3<2.0.0,>=1.26.0 in /nix/store/fbac0h6j5har9447am8mqswnf1b2mrhh-python3.10-urllib3-1.26.12/lib/python3.10/site-packages (from poetry==1.2.2) (1.26.12)
       > ERROR: Could not find a version that satisfies the requirement cleo<2.0.0,>=1.0.0a5 (from poetry) (from versions: none)
       > ERROR: No matching distribution found for cleo<2.0.0,>=1.0.0a5
       >
       For full logs, run 'nix log /nix/store/10psvwb5fqcmcdnwfnx2h64jpn5xsbfi-python3.10-poetry-1.2.2.drv'.
```

### Notify maintainers

@jbaum98 

### Metadata
Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
❯ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 5.15.79, NixOS, 23.05 (Stoat), 23.05pre431646.61a8a98e6d5`
 - - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.11.1`
 - channels(root): `"home-manager, nixos, sops-nix"`
 - channels(raito): `"home-manager, nixgl, nixpkgs"`
 - nixpkgs: `/home/raito/.nix-defexpr/channels/nixpkgs`
```

**Comment by sumnerevans** [comment]
Relevant discussion about the cleo package which broke Poetry: https://github.com/NixOS/nixpkgs/pull/203396

**Comment by tjni** [comment]
The `poetry.lock` file contains a hash of itself that makes it impossible to patch using `fetchpatch` to pull in the necessary changes. I opened https://github.com/python-poetry/poetry/discussions/7142 to discuss it.

I use this package locally to manage a non-nix Python project just to provide this as a dev dependency, so I'm personally impacted and invested in fixing this :)

In this specific case, unless there is a greater urgency, I think we can wait for a new release in https://github.com/python-poetry/poetry/pull/7137.

**Comment by m1cr0man** [comment]
Hey folks just came across this issue myself but I'd like to add something: Is it expected that `pkgs.poetry` and `pkgs.python3Packages.poetry` are different packages?

The former is 1.3.0 and appears to be defined here: https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/development/tools/poetry2nix/poetry2nix/pkgs/poetry/default.nix#L50 and builds fine.

The latter is 1.2.2 and defined here: https://github.com/NixOS/nixpkgs/blob/nixos-unstable/pkgs/development/python-modules/poetry/default.nix#L147 and suffers from the issue described here.

**Comment by tjni** [comment]
@adisbladis and @jbaum98 are the maintainers of `pkgs.poetry`, so I've CCed them for their insight.

**Comment by sersorrel** [comment]
Will this be fixed on stable once #206127 lands? (I realise that "`poetry` broken on stable" is distinct in two ways from "`python310Packages.poetry` broken on unstable" that this issue is about, but I couldn't find any other relevant issue with discussion...)

**Comment by dotlambda** [comment]
> Will this be fixed on stable once #206127 lands?

No, unrelated as you observed correctly.
This issue will be fixed once https://github.com/NixOS/nixpkgs/pull/205370 lands in master.

> I realise that "`poetry` broken on stable" is distinct in two ways from "`python310Packages.poetry` broken on unstable" that this issue is about, but I couldn't find any other relevant issue with discussion...

Yes, https://github.com/NixOS/nixpkgs/pull/206127 will fix `pkgs.poetry` in 22.11.
You could open a new issue or comment on that PR.

**Comment by DeeUnderscore** [comment]
Closing this since `poetry` now builds on unstable
